### ISSUE TITLE
Automatic nonlinear transformation of input space

### ISSUE THREAD

**Issue by stefanocereda** [scikit-optimize/scikit-optimize#506]
So far the input space is normalized into the unit hypercube in a linear way, or following a logarithm is specified by the user.
It would be nice to have an automatic decision on how to optimally wrap the search space. This is done, for example, in the paper "Input Warping for Bayesian Optimization of Non-Stationary Functions" (http://proceedings.mlr.press/v32/snoek14.html) and is (I think) the default in the Spearmint program.

It would be nice to add this functionality to scikit-optimize.
If we consider the code of Spearmint (https://github.com/HIPS/Spearmint/blob/master/spearmint/transformations/beta_warp.py) the implementation seems pretty straightforward: they just define an Alpha and Beta hyperparameters and use them to compute the forward and backward transformation.

What is blocking me is how to tune those Alpha and Beta. From the paper they say that they use a prior distribution for those parameters, and then they can be integrated out. I do not know how to implement this step in scikit-optimize. From my understanding this should be similar to what is done with the kernel's length scale, so if anybody could give me an hint it would be much appreciated.

**Comment by stefanocereda** [comment]
I've found the code for the optimization of the kernel's hyperparameters: this is actually done in scikit-learn (https://github.com/scikit-learn/scikit-learn/blob/ef5cb84a/sklearn/gaussian_process/gpr.py#L203)

I'd like to use a similar approach to tune the hyperparameters of the input space normalization, but I would also like to do this in an extensible way, so that in the future we can optimize any kind of hyperparameter. This would greatly simplify the usage of scikit-optimize in real world scenarios.

Does anybody have some ideas on how it would be better to perform this?

**Comment by MechCoder** [comment]
Look at the discussion here (https://github.com/scikit-optimize/scikit-optimize/issues/434)

TL;DR Including such a mapping means that there is no closed-form solution for the derivative of the log-likelihood with respect to the kernel hyperparameters. So, one way to do this is simply sample from posterior of the kernel hyperparameters P(theta | data) instead of finding the argmax of log(P(data | theta))
which is implemented in the original spearmint paper.

**Comment by stefanocereda** [comment]
Quick update: I have a working implementation of this and preliminary tests show really great results. I'm actually doing this in a slightly different way wrt spearmint, the optimization is performed by gradient descent instead of mcmc.
I should refine a little bit the implementation before pushing it, but I really think that we should come up with a solid solution able to handle every kind of hyperparameter.

**Comment by betatim** [comment]
Looking forward to a PR! Especially to see how this interacts/works when you use a tree based minimizer instead of a GP :)

**Comment by stefanocereda** [comment]
It won't work :(
I've done this by modifying the gp_regressor defined in scikit-learn, so it will not affect other regressors. That's one of the reasons for why I'm thinking about a cleaner solution ;)

**Comment by stefanocereda** [comment]
I've completed the testing of this feature.
I've defined a
`def branin_exp(x):
    return(branin(np.log(x[0]), np.log(x[1])))
`
and optimized on `np.exp( [(-5.0, 10.0), (0.0, 15.0)] )`

These are the results of 100 epochs of training averaged on 50 runs. In order to obtain good results with this I've used a really higher n_restarts_optimizer (500)

![image](https://user-images.githubusercontent.com/11663239/30747131-95e6aada-9fac-11e7-8558-1c11ff1e701f.png)


Moreover, this approach remains good on spaces where warping is not needed. These are the results on the normal branin:
![image](https://user-images.githubusercontent.com/11663239/30747214-c78569be-9fac-11e7-9e46-93644a8ec52a.png)

**Comment by betatim** [comment]
Have to do some more thinking/reading about the input warping so maybe my next comment is irrelevant: Implementation wise what do you think about using something along the lines of the transformers we have that implement normalization/log scaling?

**Comment by stefanocereda** [comment]
That was my first idea and is also the way in which it's implemented in Spearmint. The problem is that we need to "tune" the hyperparameters of the warper inside the fit function of the regressor (the one defined by scikit-optimize), which does not know anything about the transformers.

The best solution would be to implement it as a transformer and then devise a way to tune the transformers inside the fit function (like it's done for the kernel).

Another solution could be to implement it as a kernel, but I don't know if that would be feasible.

**Comment by betatim** [comment]
Ok, like you I also don't know what the best option is :-/ Let's see if someone else has a smart idea, as I think we will have this problem again in related things so worth taking a bit of time to think of a nice solution.

cc @glouppe @iaroslav-ai

**Comment by iaroslav-ai** [comment]
First of all, thanks for looking into this @stefanocereda - results look promising!
To me this sounds like using something like a Pipeline class, where you have a transformer and estimator, and hyperparameters of transformer are optimized by somehting using GridSearchCV. But it would be best if I could look into the code. Do you have your implementation somewhere public? So far did not found it - would love to take a look (it is also ok if you do not have it very presentable, I will find my way:) )
